### DISCUSSION TITLE
Integrate MainAnnotation with scala-cli scripts

### DISCUSSION THREAD

**Discussion by anatoliykmetyuk** [VirtusLab/scala-cli#864] [Ideas]
scala-cli can be used to execute scripts. A CLI script file, `*.sc`, performs the same role as the `@main` annotation in a normal Scala program: it marks a block of code as a runnable program.

There is a pending [PR](https://github.com/lampepfl/dotty/pull/14558) to Dotty that aims to allow users to define their own annotations similar to `@main` ([docs in PR](https://github.com/lampepfl/dotty/pull/14558/files#diff-659bde3336cdb780a624fb90a3f5ca80ea27a6d1dcd78506b5c96d51480751cb)). This mechanism can be used e.g. by framework authors to provide their users with a standard way to run the program based on their library. For example, in [Cask](https://github.com/com-lihaoyi/cask), to define an application you need to extend `cask.MainRoutes` class.

It would be good to also allow scala-cli scripts to support the customization mechanism like that. For example, take the following file from Cask documentation:

```scala
object MinimalApplication extends cask.MainRoutes{
  @cask.get("/")
  def hello() = {
    "Hello World!"
  }

  @cask.post("/do-thing")
  def doThing(request: cask.Request) = {
    request.text().reverse
  }

  initialize()
}
```

I propose to be able to write it as a Scala CLI script as follows:

```scala
//> using Scala 3.1.2
//> MainAnnotation: com.haoyi.cask.CustomMainAnnotationToBeImplemented   // or another similar way to configure it

@cask.get("/")
def hello() = {
  "Hello World!"
}

@cask.post("/do-thing")
def doThing(request: cask.Request) = {
  request.text().reverse
}
```

Notice how the class definition and the `initialize()` call go away – abstracted to the referenced macro.

The impact of the above is a simplified way to get started not only with basic scripts but more complex programs like a web server.

**Comment by bjornregnell** [comment]
Look's like a useful feature, and in the Scala spirit of generalizing concepts and allowing libraries to enable features. Nice! :)

**Comment by kubukoz** [comment]
Could this mechanism also be used for cats-effect's IOApp?

In `.scala`:

```scala
//foo.scala
import cats.effect._
object Main extends IOApp.Simple {
  override def run: IO[Unit] = IO.println("foo")
}
```

In `.sc`:

```scala
//foo.sc
//> MainAnnotation: cats.effect.IOApp

IO.println("foo")
```
